perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/VosManagerBlImpl.java

```diff
@@ -385,6 +408,76 @@ public Vo getVoById(PerunSession sess, int id) throws InternalErrorException, Vo
 		}
 	}
 
+	@Override
+	public List<MemberCandidate> getCompleteCandidates(PerunSession sess, Vo vo, List<String> attrNames, String searchString) throws InternalErrorException {
+		List<ExtSource> extSources = getPerunBl().getExtSourcesManagerBl().getVoExtSources(sess, vo);
+
+		return this.getMemberCandidates(sess, vo, null, attrNames, searchString, extSources);
+	}
+
+	@Override
+	public List<MemberCandidate> getCompleteCandidates(PerunSession sess, Vo vo, Group group, List<String> attrNames, String searchString, List<ExtSource> extSources) throws InternalErrorException {
+
+		return this.getMemberCandidates(sess, vo, group, attrNames, searchString, extSources);
+	}
+
+	/**
+	 * Finds MemberCandidates who can join the Group or Vo if the Group is null. If the given vo vo to be used is null,
+	 * the users are searched in whole perun. Otherwise, only in this vo. Group and vo can not be null at the same time.
+	 *
+	 * @param sess session
+	 * @param vo vo to be used, if is null, the users are searched in whole perun. Otherwise, only in this vo.
+	 * @param group group to be used
+	 * @param attrNames attrNames
+	 * @param searchString depends on the extSource of the Vo, could by part of the name, email or something like that.
+	 * @param extSources extSources used to find candidates
+	 * @return list of memberCandidates who match the searchString
+	 * @throws InternalErrorException internal error
+	 */
+	private List<MemberCandidate> getMemberCandidates(PerunSession sess, Vo vo, Group group, List<String> attrNames, String searchString, List<ExtSource> extSources) throws InternalErrorException {
+		List<Candidate> candidates;
+		List<RichUser> richUsers;
+
+		if (vo == null && group == null) {
+			throw new InternalErrorException("Vo and group can not be both null");
+		}
+
+		if (vo != null) {
```

Comment:
This should be ``group != null`` and then you can merge below IF on group inside this one. Right now it doesn't work correctly for vo_manager (it searches members of VO instead of whole Perun).

Reply:
Looking at he code again, I think that it is ok. When the method in entry is called by vo_admin, then the vo is set to null. So then in the getMemberCandidates() method is called the 'else' branch on line 455 and all users from perun are searched.

Reply:
On the other hand, the javadoc is not written very well, I will change it so it is easier to understand the methor.

Reply:
Fair point, but this apply only, when you use ``group,searchString,attrNames`` in RPC API, but when VO manager tries to add member to VO, ``vo,searchString,attrNames`` are used in RPC API, so you get non-null VO with VO ext sources here and then it searches for users only inside VO, which is pointless, since admin wants to add new people to the VO.

Then you should fix ``getCompleteCandidates(PerunSession sess, Vo vo, List<String> attrNames, String searchString)`` to not pass VO, but it would end-up with both objects null, which is not allowed.

When I think about it, my fix wasn't ok either, since it always returned only vo members when adding to group. Maybe you should just use some boolean flag instead of null/non-null in vo and group params.

Reply:
Yeah, you are right. I will probably change the whole logic of it, maybe with those boolean flags. I just dont like methods with a lot of arguments but on the other hand a don't want to create duplicate code.

Reply:
Ok, I have changed it. I think it is better now.